### DISCUSSION TITLE
Restart error- undefined value found for forcing variable rivsto in Noah-MP3.6

### DISCUSSION THREAD

**Discussion by Gayathridevi123** [NASA-LIS/LISF#972] [Q&A] [answered]
Hi LIS  community,

I'm trying to do a spin-up using NOAH-MP-3.6 LSM with GDAS forcings (using 2014 data) and facing two issues.

1)  First run the model as coldstart and generated the restart file. But for the second run when giving the restart file getting error like 

[INFO] GDAS -- changing grid to 2010 - 2015 (T574) --
 MSG: The GDAS forcing resolution is finer than the running domain.
      Upscaling with the average method.
 [INFO] Getting new time1 data
 [INFO] Reading GDAS file1 (I) ./input/MET_FORCING/GDAS//201401/2014010100.gdas1.sfluxgrbf00.sg
 [INFO] Reading GDAS file1 (A) ./input/MET_FORCING/GDAS//201312/2013123118.gdas1.sfluxgrbf03.sg
 [INFO] Reading GDAS file1 (A)./input/MET_FORCING/GDAS//201312/2013123118.gdas1.sfluxgrbf06.sg
 [INFO] Getting new time2 data
 [INFO] Reading GDAS file2 (I) ./input/MET_FORCING/GDAS//201401/2014010100.gdas1.sfluxgrbf03.sg
 [INFO] Reading GDAS file2 (A) ./input/MET_FORCING/GDAS//201401/2014010100.gdas1.sfluxgrbf03.sg
 undefined value found for forcing variable rivsto in NoahMP36
 for tile          115 latitude =    13.8750000     longitude =    79.6250000    
 [ERR] endrun is being called

Then I came to know when running the LIS "restart", coldstart routine for NOAH-MP-3.6 is never called. So edited the NoahMP36_readrst.F90 according to someone who had faced the same issue before. Added these lines to end of NoahMP36_readrst.F90 file,  starting at line 336: 

        do t=1,LIS_rc%npatch(n,LIS_rc%lsm_index)
                     NOAHMP36_struc(n)%noahmp36(t)%rivsto=0.0
                     NOAHMP36_struc(n)%noahmp36(t)%fldsto=0.0
                     NOAHMP36_struc(n)%noahmp36(t)%fldfrc=0.0
           enddo
    
Then also facing the same problem and I'm using LISF 7.3.1 version. I'm attaching the lislog, lis config file and the NoahMP36_readrst.F90 file.

[lislog.docx](https://github.com/NASA-LIS/LISF/files/7957462/lislog.docx)
[lis.config_noahmp.docx](https://github.com/NASA-LIS/LISF/files/7957475/lis.config_noahmp.docx)
[NoahMP36_readrst.docx](https://github.com/NASA-LIS/LISF/files/7957485/NoahMP36_readrst.docx)

2) I'm not able to generate outputs for the variables evaporative fraction(EF), Potential Evapotranspiration and Root zone soil moisture. When selecting these variables to get output, NOAH-MP-3.6 model won't run using coldstart also and giving error like "EF field is not defined for diagnostic output. Please exclude it from the model output attributes table." 

I would be extremely grateful if someone could help me with this. Hope this also helps someone facing the same issues.

Thank you.
With warm regards,
Gayathri

**Comment by bmcandr** [comment]
Hi @Gayathridevi123,

Thank you for opening this discussion. Sorry to hear you're encountering some issues.

1. @agetiran, do you have an idea about what may be causing this error? Looks like they are using LISF v7.3.1-public which does not contain the [upgrades to surface water routines](https://github.com/NASA-LIS/LISF/pull/762/files?show-viewed-files=true&file-filters%5B%5D=) that were made as part of the EIS project. Those upgrades touched the `rivsto`, `fldsto`, and `fldfrc` variables in NoahMP.3.6. ~~Perhaps upgrading to 7.3.2 or later will resolve this?~~ Correction: this fix is available in the master branch and will be available in a forthcoming public-7.4 release.

2. It may be that NoahMP.3.6 does not support outputting these variables. I do not see them among the calls to `LIS_diagnoseSurfaceOutputVar` in `lis/surfacemodels/land/noahmp.3.6/NoahMP36_main.F90` (i.e., search for `LIS_MOC_EF`). @dmocko am I looking in the right place?
    Per Appendix A in the LIS User's Guide:

    > 1 . Routine to diagnose error (variable)
    >
    >```Error:\<var\> field is not defined```
    >```for diagnostic output ..```
    >```Please exclude it from the model output attributes table```
    >```Program stopping..```
    >
    >This error occurs because the variable specified in the model output attributes file is not among the ones supported for output from the particular model. Either exclude the variable (turn it off) from the model output attributes file or implement the routine within the model to support it.

**Reply by dmocko** [reply]
> 1. @agetiran, do you have an idea about what may be causing this error? Looks like they are using LISF v7.3.1-public which does not contain the [upgrades to surface water routines](https://github.com/NASA-LIS/LISF/pull/762/files?show-viewed-files=true&file-filters%5B%5D=) that were made as part of the EIS project. Those upgrades touched the `rivsto`, `fldsto`, and `fldfrc` variables in NoahMP.3.6. Perhaps upgrading to 7.3.2 or later will resolve this?

Hi @bmcandr - Looks like we were typing our responses at the same time.  Good thing our answers are mostly consistent.

I did take a look at the latest public release (7.3.3) and this code is not yet in there.  I believe that a public-7.4 release is coming soon, which will have this fix.

David

**Reply by bmcandr** [reply]
@dmocko, thanks for the correction. I must have been looking at the wrong branch.

**Reply by agetiran** [reply]
Hi @Gayathridevi123,
The error message is related to the 2-way coupling. It looks like the version that you are using doesn't contain a fix for a case where the HyMAP routing model isn’t used. When that’s the case, those three variables (i.e., rivsto, fldsto and fldfrc) don’t get any values and the checks in NoahMP36_main.F90 (see lines below) end the run. The proposed solution is fine, but it won’t work with the 2-way coupling option, as it zeroes the variables when LIS is restarted.

            !ag (23Sep2019)
            ! check validity of rivsto
            if(tmp_rivsto .eq. LIS_rc%udef) then
                write(LIS_logunit, *) "undefined value found for forcing variable rivsto in NoahMP36"
                write(LIS_logunit, *) "for tile ", t, "latitude = ", lat, "longitude = ", lon
                call LIS_endrun()
            endif
            ! check validity of fldsto
            if(tmp_fldsto .eq. LIS_rc%udef) then
                write(LIS_logunit, *) "undefined value found for forcing variable fldsto in NoahMP36"
                write(LIS_logunit, *) "for tile ", t, "latitude = ", lat, "longitude = ", lon
                call LIS_endrun()
            endif
            ! check validity of fldfrc
            if(tmp_fldfrc .eq. LIS_rc%udef) then
                write(LIS_logunit, *) "undefined value found for forcing variable fldfrc in NoahMP36"
                write(LIS_logunit, *) "for tile ", t, "latitude = ", lat, "longitude = ", lon
                call LIS_endrun()
            endif

**Comment by dmocko** [comment, marked as answer]
Hi @Gayathridevi123 

Thanks for posting on our Discussions board.

Regarding "1", thanks for the info and for the updated code.  Yes, we had encountered this before, but it does not appear as if the updated code has made it into the public branch.  It is in the master branch, so it will be in the next public release, which we expect to happen soon.  The master branch moves these variable initializations into the lsmMod.F90 code, so it happens for both restart and coldstart mode:

https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/noahmp.3.6/NoahMP36_lsmMod.F90#L325

I'm not clear from your post, but it seems to me that your code fix resolved your question #1 - and you are just letting us know that the code is needed.  If this does not fix your first problem, please let us know.

Regarding "2", these three variables are not currently available to output in Noah-MP.3.6.

Noah-MP actually does not calculate the Potential Evapotranspiration as part of its physics calculations, so it's not possible to have the PotEvap be output from this LSM.

The Evaporative Fraction and the Root Zone Soil Moisture could be added to the output by adding some code to NoahMP36_main.F90.  You can use some code in noah.3.9/noah39_main.F90 as a guide to calculating these outputs:

        !Evaporative Fraction
        if ((eta+shtflx).ne.0) then 
           call LIS_diagnoseSurfaceOutputVar(n, t,LIS_MOC_EF,                    &
                value=abs(eta/(eta+shtflx)),vlevel=1,unit="-",&
                direction="-",surface_type=LIS_rc%lsm_index)
        else
           !double check
           call LIS_diagnoseSurfaceOutputVar(n, t,LIS_MOC_EF,value=1.0,          &
                vlevel=1,unit="-",direction="-",surface_type=LIS_rc%lsm_index)
        endif


        ! ROOT ZONE COLUMN SOIL MOISTURE IN METERS (SOILRZ)
        do k = 1,noah39_struc(n)%noah(t)%nroot
           soilrz    = soilrz + (noah39_struc(n)%noah(t)%smc(k)*sldpth(k) *    &
                                 LIS_CONST_RHOFW)
           soilrzmax = soilrzmax + (noah39_struc(n)%noah(t)%smcmax*sldpth(k) * &
                                    LIS_CONST_RHOFW) ! SY
        end do
        noah39_struc(n)%noah(t)%rootmoist = soilrz   !in volumetric
        call LIS_diagnoseSurfaceOutputVar(n,t,LIS_MOC_ROOTMOIST,vlevel=1,        &
             value=soilrz,unit="kg m-2",direction="-",surface_type=LIS_rc%lsm_index)
        call LIS_diagnoseSurfaceOutputVar(n,t,LIS_MOC_ROOTMOIST,vlevel=1,        &
             value=soilrz/1000.0,unit="m^3 m-3",direction="-",surface_type=LIS_rc%lsm_index)

You will need to change from noah39_struc%noah to NOAHMP401_struc%noahmp401, as well as ensure the correct local variable names are set and initialized for eta, shtflx, and soilrz.

David

**Reply by Gayathridevi123** [reply]
Hi  Brendan@bmcandr, David@dmocko, &  Augusto@agetiran,

Thank you so much everyone for the quick response and valuable suggestions. It really helped me.

Regarding 1. I edited that lsmMod.F90 code according to the master branch code as pointed out by David
(https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/noahmp.3.6/NoahMP36_lsmMod.F90#L325)
Then also I'm facing the same issue. And I understood from @agetiran answer that the solution won’t work with the 2-way coupling option. Right?

But I tried with NOAHMP 4.0.1 and was able to solve this problem.

Regarding 2.    I'm trying to get output for Root Zone Soil Moisture and I edited the code but not working. Can you please check and let me know, what is wrong and what I should do? I edited code like this

 1.  define the variables

real            :: tmp_soilrz           ! root zone soil moisture [Kg m-2]
real            :: tmp_soilrzmax

2. ! get parameters
tmp_soilrz            = NOAHMP401_struc(n)%noahmp401(t)%nroot

3. Initialize ROOT ZONE COLUMN SOIL MOISTURE IN METERS (SOILRZ)
       SOILRZ = 0.0
       SOILRZMAX = 0.0

4. ! call model physics
tmp_soilrz            , & ! out   - root zone soil moisture [Kg m-2]

 5. ! save output variables from local variables to global variables
NOAHMP401_struc(n)%noahmp401(t)%nroot     = tmp_soilrz

6. ! ROOT ZONE COLUMN SOIL MOISTURE IN METERS (SOILRZ)
         do k = 1,NOAHMP401_struc(n)%noahmp401(t)%nroot
            soilrz    = soilrz + (NOAHMP401_struc(n)%noahmp401(t)%smc(k)*sldpth(k) *    &
                                 LIS_CONST_RHOFW)
            soilrzmax = soilrzmax + (NOAHMP401_struc(n)%noahmp401(t)%smcmax*sldpth(k) * &
                                    LIS_CONST_RHOFW) ! SY
         end do
         NOAHMP401_struc(n)%noahmp401(t)%rootmoist = soilrz   !in volumetric
         call LIS_diagnoseSurfaceOutputVar(n,t,LIS_MOC_ROOTMOIST,vlevel=1,        &
             value=soilrz,unit="kg m-2",direction="-",surface_type=LIS_rc%lsm_index)
         call LIS_diagnoseSurfaceOutputVar(n,t,LIS_MOC_ROOTMOIST,vlevel=1,        &
             value=soilrz/1000.0,unit="m^3 m-3",direction="-",surface_type=LIS_rc%lsm_index)

and I don't know what is nroot!!

Attaching edited code here:
[NoahMP401_main.odt](https://github.com/NASA-LIS/LISF/files/7978265/NoahMP401_main.odt)

Thank you for your support...
Gayathri

**Comment by Gayathridevi123** [comment]
Hi LIS community,

I was facing the same error for NOAHMP3.6 (Restart error- undefined value found for forcing variable rivsto in Noah-MP3.6) after editing the lsmMod.F90 code according to the master branch code as suggested by David (https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/noahmp.3.6/NoahMP36_lsmMod.F90#L325). 

[In my case, no routing model options were used for the simulations]

After that tried NOAHMP 4.0.1 and was able to resolve this issue. 

But I always like to look back on problems where I failed. So compared my LISF version 7.3.1 ****noahmp3.6 F90** codes** (for beginners –these codes will be there in your /lis/surfacemodels/land/noahmp.3.6 directory) with updated codes available in the Github master branch (https://github.com/NASA-LIS/LISF/tree/master/lis/surfacemodels/land/noahmp.3.6).

Updated the F90 files (**NoahMP36_lsmMod.F90, NoahMP36_main.F90, NoahMP36_module.F90, NoahMP36_readrst.F90**) according to master branch and now I’m able to run with restart files.

**_Hope this will be helpful for others..._**

Once again thank you for the support.

Kind regards,
Gayathri
